modules/seqexec/server/src/main/scala/seqexec/server/SeqTranslate.scala

```diff
@@ -342,219 +366,253 @@ object SeqTranslate {
-      instrument,
-      systems.gemsKeywordsReader,
+      kwClient,
+      systemss.gemsKeywordsReader,
       obsKReader,
       tcsKReader
     )
 
-    private def calcHeaders(
-      config: CleanConfig,
-      stepType: StepType,
-      sys: InstrumentSystem[F]
-    ): F[HeaderExtraData => List[Header[F]]] = {
-      stepType match {
-        case StepType.CelestialObject(_) | StepType.NodAndShuffle(_) =>
-            calcInstHeader(config, sys).map(h => ctx =>
-              List(commonHeaders(config, allButGaos.toList, sys)(ctx), gwsHeaders(sys), h))
-
-        case StepType.AltairObs(_)    =>
-          for {
-            gst  <- Altair.guideStarType[F](config)
-            read <- calcInstHeader(config, sys).map(h => (ctx: HeaderExtraData) =>
-                      // Order is important
-                      List(
-                        commonHeaders(config, allButGaos.toList, sys)(ctx),
-                        altairHeader(sys),
-                        altairLgsHeader(gst, sys),
-                        gwsHeaders(sys), h))
-          } yield read
-
-        case StepType.FlatOrArc(inst) =>
-            calcInstHeader(config, sys).map(h => ctx =>
-              List(commonHeaders(config, flatOrArcTcsSubsystems(inst).toList, sys)(ctx), gcalHeader(sys), gwsHeaders(sys), h))
-
-        case StepType.NightFlatOrArc(_) =>
-            calcInstHeader(config, sys).map(h => ctx =>
-              List(commonHeaders(config, List(AGUnit, OIWFS, M2, M1, Mount), sys)(ctx), gcalHeader(sys), gwsHeaders(sys), h))
-
-        case StepType.DarkOrBias(_) | StepType.DarkOrBiasNS(_) | StepType.ExclusiveDarkOrBias(_) =>
-            calcInstHeader(config, sys).map(h => (ctx => List(commonHeaders(config, Nil, sys)(ctx), gwsHeaders(sys), h)))
-
-        case StepType.AlignAndCalib   => ((_: HeaderExtraData) => List.empty[Header[F]]).pure[F] // No headers for A&C
-
-        case StepType.Gems(_)         =>
-          val obsKReader = ObsKeywordReader[F](config, site)
-          calcInstHeader(config, sys).map(h => ctx =>
-            List(commonHeaders(config, allButGaos.toList, sys)(ctx),
-              gwsHeaders(sys),
-              gemsHeaders(sys, obsKReader, systems.tcsKeywordReader), h
-            )
+    private def calcHeaders(config: CleanConfig,
+                            stepType: StepType,
+                            instrument: Instrument
+    ): F[KeywordsClient[F] => HeaderExtraData => List[Header[F]]] = stepType match {
+      case StepType.CelestialObject(_) | StepType.NodAndShuffle(_) => { kwClient: KeywordsClient[F] => ctx: HeaderExtraData =>
+        List(
+          commonHeaders(config, allButGaos.toList, kwClient)(ctx),
+          gwsHeaders(kwClient),
+          calcInstHeader(config, instrument, kwClient)
+        )
+      }.pure[F]
+
+      case StepType.AltairObs(_)    => Altair.guideStarType[F](config).map { gst =>
+        kwClient: KeywordsClient[F] => ctx: HeaderExtraData =>
+          // Order is important
```

Comment:
If order is important should we use a structure that preserves order? e.g. `SortedMap`?